docs: remote deployment example (device-usb-camera) added (#1211)
edgexfoundry/edgex-docs: docs_src/microservices/device/remote/Ch-RemoteNonSecure.md

Reviewer: In step 4, I was updating docker-compose-base.yml file, and also was changing ip address for database, consul and core-metadata not for all edgex core services

Reviewer 2: Nothing should change IP wise in the core service compose file. IP changes are only needed in the remote service.

# Remote deployment of device services in non-secure mode

> Usually microservices are deployed using multiple nodes for scalability, availability and performance 
> instead of running all services in one node.

Reviewer 2: I don't agree with `Usually`, this is about deploying the device service on the nodes where the device is connected.
```suggestion
In some use cases, devices are connected to nodes where main EdgeX services are not running. In these cases the appropriate device service needs to run on the remote nodes where it can connect to the device(s) and communicate to the main node where the rest of the EdgeX services are running.
```

# Non-Secure Mode

## Remote deployment of device services in non-secure mode
Usually microservices are deployed using multiple nodes for scalability, availability and performance 
instead of running all services in one node.

This page provides an example of remote deployment of `device-usb-camera` service using multiple nodes in non-secure mode.
The deployment can be done by running the service `natively` or by running it in `docker`.

## Example

> This example uses 2 nodes for remote deployment. One node (remote) is used to run all Edgex core services in docker and the other node (host)

Reviewer 3: ```suggestion
This example uses 2 nodes for remote deployment. One node (remote) is used to run all EdgeX core services in Docker and the other node (host)
```

> This example uses 2 nodes for remote deployment. One node (remote) is used to run all Edgex core services in docker and the other node (host)

Reviewer: this looks confusing to me. which node is host? the one running edgex or device usb?
can we say some thing like this? This example uses 2 nodes for remote deployment. One node (gateway) is used to run all EdgeX core services in Docker and the other node (remote) is used to run USB device service.

> for running the device-usb-camera service natively or in docker. Both the nodes are on the same network.

Reviewer 3: ```suggestion
for running the device-usb-camera service natively or in Docker. Both of the nodes are on the same network.
```

This example can be further expanded to run multiple instances of device-usb-camera service in multiple nodes.

## Running of the example

> 1. Set up the two nodes to be ready for remote deployment. Refer [USB Service Setup](../supported/device-usb-camera/walkthrough/setup.md)

Reviewer 3: ```suggestion
1. Set up the two nodes to be ready for remote deployment. Refer to [USB Service Setup](../supported/device-usb-camera/walkthrough/setup.md)
```

>    for system requirements and dependencies such as Git, Docker, Docker compose, etc. Additional to this Golang needs to be installed

Reviewer 3: ```suggestion
   for system requirements and dependencies such as Git, Docker, Docker Compose, etc. Additionally, Golang needs to be installed
```

   in the host node where the device-usb-camera service will be built and run natively.

> 1. Next step is to install Edgex compose in the remote node which will be used to run all Edgex core services. So clone the `edgex-compose`

Reviewer 3: ```suggestion
1. Next step is to install EdgeX Compose in the remote node which will be used to run all EdgeX core services. So clone the `edgex-compose`
```

   repository:

     ```bash
     git clone https://github.com/edgexfoundry/edgex-compose.git
     ```

1. Checkout the required version:

      ```bash
      git checkout {{version}}
      ```

> 1. Update the `docker-compose-no-secty.yml` file by changing the `host_ip` address of all the Edgex core services to the remote node ip address.

Reviewer 3: ```suggestion
1. Update the `docker-compose-no-secty.yml` file by changing the `host_ip` address of all the EdgeX core services to the remote node ip address.
```

   Non-Edgex core services such as device-rest, device-virtual, app-rules-engine, etc. can be removed or commented out if needed.

1. Run Edgex core services:

      ```bash
      make run no-secty
      ```

   1. Verify the services are up and running:

      ```bash
      docker ps 
      ```

1. Follow this guide to run the `device-usb-camera` service in docker or natively.

=== "Docker"

    Todo

=== "Native"

    1. Clone the `device-usb-camera` service repository

        ```bash
        git clone https://github.com/edgexfoundry/device-usb-camera.git
        ```
    
    1. Checkout the required version

         ```bash
         git checkout {{version}}
         ```

    1. cd into `/cmd/res` directory
         
         ```bash
         cd /cmd/res
         ```

    1. Modify the `configuration.yml` file in the repo to run natively. Uncomment out the required parts as directed in the file.

>        Update the yml file with the `Host` and `Remote` node ip addresses.All the 

Reviewer 3: ```suggestion
       Update the yml file with the `Host` and `Remote` node ip addresses. All the 
```

## Example
This example uses 2 nodes for remote deployment. One of the nodes (host) is used to run all EdgeX core services in Docker and the other node (remote)
for running the device-usb-camera service either natively or in Docker. Both the nodes are on the same network.
This example can be further extended to run multiple instances of device-usb-camera service in multiple nodes.

## Running of the example

1. Set up the two nodes to be ready for remote deployment. Refer to [USB Service Setup](../services/device-usb-camera/walkthrough/setup.md)
   for system requirements and dependencies such as Git, Docker, Docker Compose, etc. Additionally Golang needs to be installed
   in the remote node if the device-usb-camera service will be built and run natively.

1. Next step is to install [EdgeX compose](https://github.com/edgexfoundry/edgex-compose) in the host node which will be used to run all EdgeX core services. So clone the `edgex-compose`
   repository:

     ```bash
     git clone https://github.com/edgexfoundry/edgex-compose.git
     ```

1. Checkout the required version:

      ```bash
      git checkout {{edgexversion}}
      ```

> 1. Update the [docker-compose-no-secty.yml](https://github.com/edgexfoundry/edgex-compose/blob/main/docker-compose-no-secty.yml) file by removing the `host_ip` section of all the Edgex core services. E.g.

Reviewer 3: ```suggestion
1. Update the [docker-compose-no-secty.yml](https://github.com/edgexfoundry/edgex-compose/blob/main/docker-compose-no-secty.yml) file by removing the `host_ip` section of all the EdgeX core services. E.g.
```

      ```bash
      host_ip: 127.0.0.1
      ```
   The example line provided above should be removed from the services, the host_ip will be provided while running the usb service.

>    Non-Edgex core services such as device-rest, device-virtual, app-rules-engine, etc. can be removed or commented out if needed.

Reviewer 3: ```suggestion
   Non-EdgeX core services such as device-rest, device-virtual, app-rules-engine, etc. can be removed or commented out if needed.
```

> 1. Run Edgex core services:

Reviewer 3: ```suggestion
1. Run EdgeX core services:
```

      ```bash
      make run no-secty
      ```

   1. Verify the services are up and running:

      ```bash
      docker ps 
      ```

   1. Clone the [device-usb-camera](https://github.com/edgexfoundry/device-usb-camera) service repository:

       ```bash
       git clone https://github.com/edgexfoundry/device-usb-camera.git
       ```

   1. Checkout the required version:

       ```bash
       git checkout {{edgexversion}}
       ```

>    1. Follow the guide below to build and run the `device-usb-camera` service in Docker or natively.

Reviewer 2: These steps need to be under new parent topic, not `Run EdgeX core services`

=== "Docker"

    1. Build the service from the `main` directory:

         ```bash
         make docker
         ```

    1. Create `docker-compose.yml` file from anywhere in the remote node to run the device service in Docker. Copy the content below into the compose file and edit with approriate values:

        ```bash
        name: edgex
        services:
            device-usb-camera:
            container_name: edgex-device-usb-camera
            device_cgroup_rules:
            - c 81:* rw
            environment:
              EDGEX_SECURITY_SECRET_STORE: "false"

>               EDGEX_REMOTE_SERVICE_HOSTS: "<host-ip-address>,<remote-ip-address>,<service-bind-address>"

Reviewer 2: These names are confusing on which system they refer to. How about, which is consistent with non-docker example below.
```suggestion
              EDGEX_REMOTE_SERVICE_HOSTS: "<remote-node-ip-address>,<host-node-ip-address>,<service-bind-address>"
```

              #E.g.EDGEX_REMOTE_SERVICE_HOSTS: "172.118.1.92,172.118.1.167,0.0.0.0"
            hostname: edgex-device-usb-camera

>             image: <device-usb-camera-docker-image previously built>

Reviewer 2: In docker there is no reason to build the image locally. Use the published image
```suggestion
            image: <published image for device-usb-camera>
            #E.g.  image: nexus3.edgexfoundry.org:10004/device-usb-camera:latest or
            #        image: edgexfoundry/device-usb-camera:{{edgexversion}}.0
```

            ports:
            - "59983:59983"
            - "8554:8554"

>             - "8000:8000"

Reviewer 2: Why is this port here?

Author: I think this port is needed for RTSP Authentication -  https://github.com/edgexfoundry/device-usb-camera/blob/main/cmd/res/configuration.yaml#L46

Reviewer 2: It isn't here in compose builder: https://github.com/edgexfoundry/edgex-compose/blob/main/compose-builder/add-device-usb-camera.yml#L22-L24

 Should it be added to compose builder?

Author: This port only needs to be added if the RTSP server is not part of same container or network which is not the case with the default installation. So I have removed it.

            read_only: true
            restart: always
            security_opt:
            - no-new-privileges:true
            user: root:root
            volumes:
            - type: bind
              source: /dev
              target: /dev
              bind:
                create_host_path: true
            - type: bind
              source: /run/udev
              target: /run/udev
              read_only: true
              bind:
                create_host_path: true
        ```
        
        !!! note

>             If multiple instances of the service has to be run then add `EDGEX_INSTANCE_NAME` enviroment variable above with a value of number of instances desired.

Reviewer 3: ```suggestion
            If multiple instances of the service have to be run, then add `EDGEX_INSTANCE_NAME` environment variable above with a number to instantiate the current instance of the device-usb-camera service.
```

      
         ```bash
         docker compose up -d
         ```

=== "Native"

    1. For RTSP streaming get [rtsp-simple-server](https://github.com/bluenviron/mediamtx/releases) binary
       and rtsp config yml file and copy them into the [cmd](https://github.com/edgexfoundry/device-usb-camera/tree/main/cmd) directory.

    1. Build the service from the `main` directory:

         ```bash
         make build
         ```

    1. Run the service. Use appropriate ip addresses for the `rsh` flag parameter. Refer to [Remote Service Hosts](../../configuration/CommonCommandLineOptions.md#remote-service-hosts)
       for more info.

         ```bash
         EDGEX_SECURITY_SECRET_STORE=false ./cmd/device-usb-camera -cp -r -rsh=<remote-node-ip-address>,<host-node-ip-address>,<service-bind-address>
         ```

        !!! note
            If multiple instances of the service has to be run then attach `-i` followed by the number of instances desired. E.g:
            ```bash
            EDGEX_SECURITY_SECRET_STORE=false ./cmd/device-usb-camera -cp -r -rsh=172.26.113.174,172.26.113.150,0.0.0.0 -i 2
            ```

## Verify Service, Device(s) and next steps

> 1. Make sure the service has no errors and check whether the service is added to Edgex i.e. to core-metadata (running in host node):

Reviewer 3: ```suggestion
1. Make sure the service has no errors and check whether the service is added to EdgeX i.e. to core-metadata (running in host node):
```

    ```bash

>     curl -s http://<core-metadata-ip-address>:59881/api/{{api_version}}/deviceservice/name/device-usb-camera | jq .

Reviewer 2: ```suggestion
    curl -s http://<host-node-ip-address>:59881/api/{{api_version}}/deviceservice/name/device-usb-camera | jq .
```

    ```

    Successful:
    ```json
    {
        "apiVersion" : "{{api_version}}",
        "statusCode": 200,
        "service": {
            "created": 1658769423192,
            "modified": 1658872893286,
            "id": "04470def-7b5b-4362-9958-bc5ff9f54f1e",
            "name": "device-usb-camera",
            "baseAddress": "http://edgex-device-usb-camera:59983",
            "adminState": "UNLOCKED"
        }
    }
    ```
    Unsuccessful:
    ```json
    {
        "apiVersion" : "{{api_version}}",
        "message": "fail to query device service by name device-usb-camera",
        "statusCode": 404
    }
    ```       

1. Verify device(s) have been successfully added:

    ```bash

>     curl -s http://<core-metadata-ip-address>:59881/api/{{api_version}}/device/all | jq -r '"deviceName: " + '.devices[].name''

Reviewer 2: ```suggestion
    curl -s http://<host-node-ip-address>:59881/api/{{api_version}}/device/all | jq -r '"deviceName: " + '.devices[].name''
```
